src/agent/advertising_proxy.cpp

```diff
@@ -0,0 +1,316 @@
+    {
+        update.mCount += !otSrpServerServiceIsDeleted(service);
+    }
+    mOutstandingUpdates.emplace_back(update);
+
+    SuccessOrExit(error = SplitFullHostName(otSrpServerHostGetFullName(aHost), hostName, hostDomain));
+
+    if (!hostDeleted)
+    {
+        // TODO: select a preferred address.
+        otbrLog(OTBR_LOG_INFO, "[adproxy] publish SRP host: %s", hostName.c_str());
+        SuccessOrExit(error =
+                          mPublisher->PublishHost(hostName.c_str(), hostAddress[0].mFields.m8, sizeof(hostAddress[0])));
+    }
+    else
+    {
+        otbrLog(OTBR_LOG_INFO, "[adproxy] unpublish SRP host: %s", hostName.c_str());
+        SuccessOrExit(error = mPublisher->UnpublishHost(hostName.c_str()));
+    }
+
+    service = nullptr;
+    while ((service = otSrpServerHostGetNextService(aHost, service)) != nullptr)
+    {
+        const char *fullName = otSrpServerServiceGetFullName(service);
+        std::string serviceName;
+        std::string serviceType;
+        std::string serviceDomain;
+
+        SuccessOrExit(error = SplitFullServiceName(fullName, serviceName, serviceType, serviceDomain));
+
+        if (!hostDeleted && !otSrpServerServiceIsDeleted(service))
+        {
+            Mdns::Publisher::TxtList txtList = MakeTxtList(service);
+
+            otbrLog(OTBR_LOG_INFO, "[adproxy] publish SRP service: %s", fullName);
+            SuccessOrExit(error = mPublisher->PublishService(hostName.c_str(), otSrpServerServiceGetPort(service),
+                                                             serviceName.c_str(), serviceType.c_str(), txtList));
+        }
+        else
+        {
+            otbrLog(OTBR_LOG_INFO, "[adproxy] unpublish SRP service: %s", fullName);
+            SuccessOrExit(error = mPublisher->UnpublishService(serviceName.c_str(), serviceType.c_str()));
+        }
+    }
+
+exit:
+    if (error != OTBR_ERROR_NONE || update.mCount == 0)
+    {
+        if (error != OTBR_ERROR_NONE)
+        {
+            otbrLog(OTBR_LOG_INFO, "[adproxy] failed to advertise SRP service updates %p", aHost);
+        }
+
+        mOutstandingUpdates.pop_back();
+        otSrpServerHandleServiceUpdateResult(mInstance, aHost, OtbrErrorToOtError(error));
+    }
+
+    return;
+}
+
+void AdvertisingProxy::PublishServiceHandler(const char *aName, const char *aType, otbrError aError, void *aContext)
+{
+    static_cast<AdvertisingProxy *>(aContext)->PublishServiceHandler(aName, aType, aError);
+}
+
+void AdvertisingProxy::PublishServiceHandler(const char *aName, const char *aType, otbrError aError)
+{
+    otbrError error = OTBR_ERROR_NONE;
+
+    otbrLog(OTBR_LOG_INFO, "[adproxy] Handle publish service '%s' result: %d", aName, aError);
+
+    for (auto update = mOutstandingUpdates.begin(); update != mOutstandingUpdates.end(); ++update)
+    {
+        const otSrpServerService *service = nullptr;
+
+        while ((service = otSrpServerHostGetNextService(update->mHost, service)) != nullptr)
```

Comment:
Could the `Host` referenced by `mHost` be released in OT?

Reply:
updated to never reference the content host.